freeciv-web/src/main/webapp/javascript/player.js

```diff
@@ -271,3 +271,25 @@ function get_invalid_username_reason(username)
   }
   return null;
 }
+
+/**************************************************************************
+ returns the the player's capital city, or undefined
+ **************************************************************************/
+function player_capital(player)
+{
+  for (var city_id in cities) {
+    var city = cities[city_id];
+    if (does_player_own_city(player, city) && is_capital(city)) {
+      return true;
+    }
+  }
+  return false;
```

Comment:
You surely mean to return `city` and `undefined`.
Sorry I didn't spot this before.

Reply:
Indeed, though you didn't miss it: this originally used `Object.values(...)`.

Reply:
Also adjusted improvement_id_by_name to always return an int value - I realized that if a ruleset omitted a requested building, returning `undefined` or `null` could in some cases be considered the equivalent to `0` given js type-conversion.